vkdemo/plugin.cpp

```diff
@@ -0,0 +1,793 @@
+        : sb{pb->lookup_impl<switchboard>()}
+        , ds{pb->lookup_impl<display_sink>()}
+        , _m_clock{pb->lookup_impl<RelativeClock>()} { }
+
+    void initialize() {
+        if (ds->vma_allocator) {
+            this->vma_allocator = ds->vma_allocator;
+        } else {
+            this->vma_allocator = vulkan_utils::create_vma_allocator(ds->vk_instance, ds->vk_physical_device, ds->vk_device);
+        }
+
+        command_pool   = vulkan_utils::create_command_pool(ds->vk_device, ds->graphics_queue_family);
+        command_buffer = vulkan_utils::create_command_buffer(ds->vk_device, command_pool);
+        load_model();
+        bake_models();
+        create_texture_sampler();
+        create_descriptor_set_layout();
+        create_uniform_buffers();
+        create_descriptor_pool();
+        create_descriptor_set();
+        create_vertex_buffer();
+        create_index_buffer();
+        vertices.clear();
+        indices.clear();
+
+        // Construct perspective projection matrix
+        math_util::projection_fov(&basic_projection, display_params::fov_x / 2.0f, display_params::fov_x / 2.0f,
+                                  display_params::fov_y / 2.0f, display_params::fov_y / 2.0f, rendering_params::near_z,
+                                  rendering_params::far_z);
+    }
+
+    virtual void setup(VkRenderPass render_pass, uint32_t subpass) override {
+        create_pipeline(render_pass, subpass);
+    }
+
+    virtual void update_uniforms(const pose_type fp) override {
+        update_uniform(fp, 0);
+        update_uniform(fp, 1);
+    }
+
+    virtual void record_command_buffer(VkCommandBuffer commandBuffer, int eye) override {
+        vkCmdBindPipeline(commandBuffer, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline);
+        VkBuffer     vertexBuffers[] = {vertex_buffer};
+        VkDeviceSize offsets[]       = {0};
+        vkCmdBindVertexBuffers(commandBuffer, 0, 1, vertexBuffers, offsets);
+        vkCmdBindIndexBuffer(commandBuffer, index_buffer, 0, VK_INDEX_TYPE_UINT32);
+        vkCmdBindDescriptorSets(commandBuffer, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_layout, 0, 1, &descriptor_sets[eye], 0,
+                                nullptr);
+
+        for (auto& model : models) {
+            ModelPushConstant push_constant{};
+            push_constant.texture_index = texture_map[model.texture_index];
+            vkCmdPushConstants(commandBuffer, pipeline_layout, VK_SHADER_STAGE_FRAGMENT_BIT, 0, sizeof(ModelPushConstant),
+                               &push_constant);
+            vkCmdDrawIndexed(commandBuffer, model.index_count, 1, model.index_offset, 0, 0);
+        }
+    }
+
+    virtual void destroy() override { }
+
+private:
+    void update_uniform(const pose_type& fp, int eye) {
+        Eigen::Matrix4f modelMatrix = Eigen::Matrix4f::Identity();
+
+        auto pose = fp;
+
+        Eigen::Matrix3f head_rotation_matrix = pose.orientation.toRotationMatrix();
+
+        constexpr int LEFT_EYE = 0;
+
```

Comment:
Appropriate level of commenting - it would be good to have this level of comments in other parts of the code